site/components/pages/Retirements/SingleRetirement/DownloadCertificateButton/generateCertificate.ts

```diff
@@ -0,0 +1,281 @@
+import { jsPDF } from "jspdf";
+import { trimWithLocale } from "@klimadao/lib/utils";
+
+import { KlimaRetire } from "@klimadao/lib/types/subgraph";
+import { VerraProjectDetails } from "@klimadao/lib/types/verra";
+import { RetirementToken } from "@klimadao/lib/constants";
+import { urls } from "@klimadao/lib/constants";
+
+import KlimaLogo from "public/logo-klima.png";
+import bctBackground from "public/bg_bct.jpeg";
+import nctBackground from "public/bg_nct.jpeg";
+import nboBackground from "public/bg_nbo.jpeg";
+import uboBackground from "public/bg_ubo.jpeg";
+import mco2Background from "public/bg_mco2.jpeg";
+
+import { PoppinsExtraLight } from "./poppinsExtraLightbase64";
+import { PoppinsBold } from "./poppinsBoldbase64";
+
+type Params = {
+  beneficiaryName: string;
+  beneficiaryAddress: string;
+  projectDetails?: VerraProjectDetails;
+  retirement: KlimaRetire;
+  retirementMessage: string;
+  retirementUrl: string;
+  tokenData: {
+    key: string;
+    icon: StaticImageData;
+    label: Uppercase<RetirementToken>;
+  };
+};
+
+const KLIMA_GREEN = "#00cc33";
+const PRIMARY_FONT_COLOR = "#313131";
+const SECONDARY_FONT_COLOR = "#767676";
+const spacing = {
+  margin: 15,
+  mainTextWidth: 160,
+  beneficiaryName: 81,
+  transactionDetails: 130,
+  projectDetails: 152,
+  tokenImage: { x: 143, y: 156 },
+  retirementLink: 200,
+};
+
+type FeatureImageMappingKey = keyof typeof featureImageMap;
+const featureImageMap = {
+  bct: bctBackground,
+  nct: nctBackground,
+  ubo: uboBackground,
+  nbo: nboBackground,
+  mco2: mco2Background,
+};
+
+export const generateCertificate = (params: Params): void => {
+  const isMossRetirement = params.retirement.offset.bridge === "Moss";
+
+  const doc = new jsPDF({
+    orientation: "landscape",
+    format: "letter",
+    putOnlyUsedFonts: true,
+    compress: true,
+  });
+
+  const setupFonts = (): void => {
+    doc.addFileToVFS("Poppins-ExtraLight-normal.ttf", PoppinsExtraLight);
+    doc.addFileToVFS("Poppins-SemiBold-normal.ttf", PoppinsBold);
+    doc.addFont("Poppins-ExtraLight-normal.ttf", "Poppins", "ExtraLight");
+    doc.addFont("Poppins-SemiBold-normal.ttf", "Poppins", "Bold");
+  };
+
+  const printHeader = (): void => {
+    const klimaLogo = new Image();
+    klimaLogo.src = KlimaLogo.src;
+    doc.addImage(klimaLogo, "JPEG", spacing.margin, spacing.margin, 60, 10);
+
+    doc.setFont("Poppins", "Bold");
+    doc.setFontSize(24);
+    doc.text("Certificate for On-chain", spacing.margin, 36);
+    doc.text("Carbon Retirement", spacing.margin, 46);
+    doc.setLineWidth(1.05);
+    doc.setDrawColor(KLIMA_GREEN);
+    doc.line(spacing.margin, 52, 173, 52);
```

Comment:
I imagine all of this was a ton of trial and error- really appreciate it! It looks great!